components/DesktopInventoryItemList/index.tsx

```diff
@@ -172,35 +162,90 @@ interface Props {
   const [page, setPage] = React.useState(0)
-  const [rowsPerPage, setRowsPerPage] = React.useState(5)
+  const [rowsPerPage, setRowsPerPage] = React.useState(DEFAULT_ROWS_PER_PAGE)
+  const [visibleRows, setVisibleRows] = React.useState<Data[] | null>(null)
+  const [tableData, setTableData] = React.useState<Data[]>([])
 
-  const handleRequestSort = (
-    event: React.MouseEvent<unknown>,
-    property: keyof Data
-  ) => {
-    const isAsc = orderBy === property && order === 'asc'
-    setOrder(isAsc ? 'desc' : 'asc')
-    setOrderBy(property)
-  }
+  React.useEffect(() => {
+    const newTableData = props.inventoryItems.map((item) => {
+      return {
+        name: item.itemDefinition.name,
+        attributes: item.attributes,
+        category: item.itemDefinition.category?.name,
+        quantity: item.quantity,
+        assignee: item.assignee?.name,
+        kebab: '',
+        _id: item._id,
+        lowStockThreshold: item.itemDefinition.lowStockThreshold,
+        criticalStockThreshold: item.itemDefinition.criticalStockThreshold,
+        inventoryItem: item,
+      }
+    })
+    setTableData(newTableData)
+    let rowsOnMount = stableSort(
+      newTableData,
+      getComparator(DEFAULT_ORDER, DEFAULT_ORDER_BY)
+    )
+    rowsOnMount = rowsOnMount.slice(
+      0 * DEFAULT_ROWS_PER_PAGE,
+      0 * DEFAULT_ROWS_PER_PAGE + DEFAULT_ROWS_PER_PAGE
+    )
+
+    setVisibleRows(rowsOnMount)
+  }, [])
+
+  const handleRequestSort = React.useCallback(
+    (event: React.MouseEvent<unknown>, newOrderBy: keyof Data) => {
+      const isAsc = orderBy === newOrderBy && order === 'asc'
+      const toggledOrder = isAsc ? 'desc' : 'asc'
+      setOrder(toggledOrder)
+      setOrderBy(newOrderBy)
+
+      const sortedRows = stableSort(
+        tableData,
+        getComparator(toggledOrder, newOrderBy)
+      )
+      const updatedRows = sortedRows.slice(
+        page * rowsPerPage,
+        page * rowsPerPage + rowsPerPage
+      )
+      setVisibleRows(updatedRows)
+    },
+    [order, orderBy, page, rowsPerPage, tableData]
+  )
 
   const handleChangePage = (event: unknown, newPage: number) => {
     setPage(newPage)
+    const sortedRows = stableSort(tableData, getComparator(order, orderBy))
+    const updatedRows = sortedRows.slice(
+      newPage * rowsPerPage,
+      newPage * rowsPerPage + rowsPerPage
+    )
+    setVisibleRows(updatedRows)
   }
 
   const handleChangeRowsPerPage = (
     event: React.ChangeEvent<HTMLInputElement>
   ) => {
-    setRowsPerPage(parseInt(event.target.value, 10))
+    const updatedRowsPerPage = parseInt(event.target.value, 10)
+    setRowsPerPage(updatedRowsPerPage)
     setPage(0)
-  }
 
-  // Avoid a layout jump when reaching the last page with empty rows.
-  const emptyRows =
-    page > 0 ? Math.max(0, (1 + page) * rowsPerPage - rows.length) : 0
+    const sortedRows = stableSort(tableData, getComparator(order, orderBy))
+    const updatedRows = sortedRows.slice(
```

Comment:
Type of `updatedRows` incompatible with `Data[]` when used below.

The error shows each field having type `string | number` but in `Data`, most types have `string` and a few have specialized types such as the `attributes` member.